Fix state leaking when a function component throws on server render

pmaccart opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/react) and create your branch from `master`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
  5. Run `yarn test-prod` to test in the production environment. It supports the same options as `yarn test`.
  6. If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
  7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
  8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
  9. Run the [Flow](https://flowtype.org/) type checks (`yarn flow`).
  10. If you haven't already, complete the CLA.

  Learn more about contributing: https://reactjs.org/docs/how-to-contribute.html
-->

## Summary
Reset the internal hooks state when preparing a component for hooks usage, rather than after a component finishes rendering.  This fixes https://github.com/facebook/react/issues/19211, where the internal hooks state is not cleared if a function component throws an error while rendering.  

## Test Plan
- ` yarn test-prod`: validated no test regressions were introduced
- Updated the repro case from https://github.com/pmaccart/react-hooks-ssr-state-leak to use the new build; validated that the example scenario successfully rendered after set

gaearon commented:
Can you add a test that would have been failing before? That would make it easier to check the changes.

gaearon commented:
Can we instead add `resetHooksAfterThrow`, mirroring what we do in `ReactFiberHooks`?

FiberesimaJoseph commented:
Joseph Fiberesima

const Signup = () => {
    const [username, setUsername] = useState('')
    const [email, setEmail] = useState('')
    const [password, setPassword] = useState('')
    const [passwordConfirmation, setPasswordConfirmation] = useState('')

    const clearState = () => {
        setUsername('')
        setEmail('')
        setPassword('')
        setPasswordConfirmation('')
    }

    const handleSubmit = signupUser => e => {
        e.preventDefault()
        signupUser().then(data => {
            console.log(data)
            clearState() // <-----------
        })
    }

    return <JSX />
}

export default Signup
Each piece of state is used for a controlled input for the form.
.

Your code looks good, but if you want to decrease the functions needed you can put your entire form state in a single state variable object and reset to the initial object.

Example

const { useState } = React;

function signupUser() {
  return new Promise(resolve => {
    setTimeout(resolve, 1000);
  });
}

const initialState = {
  username: "",
  email: "",
  password: "",
  passwordConfirmation: ""
};

const Signup = () => {
  const [
    { username, email, password, passwordConfirmation },
    setState
  ] = useState(initialState);

  const clearState = () => {
    setState({ ...initialState });
  };

  const onChange = e => {
    const { name, value } = e.target;
    setState(prevState => ({ ...prevState, [name]: value }));
  };

  const handleSubmit = e => {
    e.preventDefault();
    signupUser().then(clearState);
  };

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label>
          Username:
          <input value={username} name="username" onChange={onChange} />
        </label>
      </div>
      <div>
        <label>
          Email:
          <input value={email} name="email" onChange={onChange} />
        </label>
      </div>
      <div>
        <label>
          Password:
          <input
            value={password}
            name="password"
            type="password"
            onChange={onChange}
          />
        </label>
      </div>
      <div>
        <label>
          Confirm Password:
          <input
            value={passwordConfirmation}
            name="passwordConfirmation"
            type="password"
            onChange={onChange}
          />
        </label>
      </div>
      <button>Submit</button>
    </form>
  );
};

ReactDOM.render(<Signup />, document.getElementById("root"));
<script src="https://unpkg.com/react@16/umd/react.development.js"></script>
<script src="https://unpkg.com/react-dom@16/umd/react-dom.development.js"></script>

<div id="root"></div>

pmaccart commented:
@gaearon Thanks for the feedback, and sorry for the slow response.

I made a couple updates to the PR:
1.  Added a unit test that fails prior to the fix ([64f35a1](https://github.com/facebook/react/pull/19212/commits/64f35a1f1966e1d850b76508c55b48ef0761ddab))
2.  Added a commit that implements a similar approach to what is taken in `ReactFiberHooks` -- eager to get your feedback on whether this approach is preferable, or whether we want to stick with resetting the state in the `prepareToUseHooks` function.  Happy to go either way on this.

gaearon commented on packages/react-dom/src/__tests__/ReactDOMServerIntegrationHooks-test.js:
```diff
@@ -867,6 +867,48 @@ describe('ReactDOMServerHooks', () => {
+
+    // First, render a component that will throw an error during a re-render triggered
+    // by a dispatch call.
+    try {
```
Can you wrap this in `expect(...).toThrow()` instead so we can assert the throw happened and the message matched what we expected?

pmaccart replied:
Ah yes -- done.

gaearon commented on packages/react-dom/src/server/ReactPartialRendererHooks.js:
```diff
@@ -202,16 +202,12 @@ export function finishHooks(
   }
 
-  // These were reset above
+  // These were reset via the resetHooksState() call
```
Let's just remove this comment?

pmaccart replied:
Done.

gaearon commented on packages/react-dom/src/server/ReactPartialRendererHooks.js:
```diff
@@ -202,16 +202,12 @@ export function finishHooks(
-  renderPhaseUpdates = null;
-  workInProgressHook = null;
+  resetHooksState();
   if (__DEV__) {
```
Move this DEV-only block to reset too?

pmaccart replied:
Done.

gaearon commented on packages/react-dom/src/server/ReactPartialRendererHooks.js:
```diff
@@ -202,24 +202,21 @@ export function finishHooks(
 
     children = Component(props, refOrContext);
   }
-  currentlyRenderingComponent = null;
```
Did we lose currentlyRenderingComponent?

pmaccart replied:
Oops, yes -- added that in as well.

gaearon commented:
Looks good. Thx
